Add instructions to modify custom Docker images (#12567)
PrairieLearn/PrairieLearn: docs/dockerImages.md

1. Install [Docker Desktop](https://docs.docker.com/desktop/) in your environment.
2. Create a [Docker Hub](https://hub.docker.com/) account, if you don't yet have one.

   !!! warning

>    This step should be done by a user that will maintain a long-term association with the course in question, such as a permanent faculty member or support staff, not a teaching assistant or sessional instructor. This will simplify any future updates to custom images, if they become necessary. If multiple users are expected to get access to push these images, you may consider the use of [access tokens](https://docs.docker.com/security/access-tokens/) or [teams](https://docs.docker.com/admin/organization/manage-a-team/).

Author: This is not causing a warning box because it is indented by 3 spaces, and the admonitions plugin [seems to require 4 spaces](https://github.com/squidfunk/mkdocs-material/discussions/6290#discussioncomment-7474767). Changing it to 4 spaces causes prettier to revert to 3. Working on a workaround.

   ```

   Note that the production environment used by PrairieLearn requires a `linux/amd64` platform for images. As such it is necessary that images are built using this platform, as listed in the `--platform` argument in the command above. This is particularly relevant for users of ARM-based architectures such as Apple Silicon or Windows on Arm systems, since these systems default to `linux/arm64` instead.

6. Once the build process completes, push the image to the Docker Hub registry (again replacing `yourdockerhubaccount` and `yourimagename` as needed):

   ```bash
   docker push yourdockerhubaccount/yourimagename
   ```

7. Change the settings for the question that needs the custom image to use the image name you set above.
8. In your course's "Sync" page, under "Docker images", find the image name above and click on "Sync".

Note that the process above _will not_ cause the image above to be automatically updated if there are changes or additional features to the original image. It is your responsibility to periodically update your custom image by repeating steps 5-8 above.

### Testing custom images locally

> For simple custom images that only include one or two extra packages or libraries, the steps above should be sufficient to ensure that the image works as expected, testing them in production itself. Similarly, simple changes like updating package versions can be achieved by updating the files from step 4 above, then re-running steps 5-8.

Reviewer: Nit, the other phrasing felt a bit clunky/unclear to me, specifically "testing them in production itself" at the end of the sentence. A potential change:

```suggestion
For simple custom images that only include one or two extra packages or libraries, following the steps above and testing the image in production may be sufficient to ensure that it works as expected. Similarly, simple changes like updating package versions can be achieved by updating the files from step 4 above, then re-running steps 5-8.
```

Feel free to iterate on this if you were trying to convey something different.

> However, for more complex images or more elaborate changes to existing custom images, it is advisable to test the image locally in a [local PrairieLearn environment](./installing.md#support-for-external-graders-and-workspaces) before pushing it to the production environment. In this scenario, it is important to ensure that the updated version of the image is restricted to the local environment, and that the production environment is not updated with the new image version until testing is complete. While avoiding an image sync (step 8 above) is often enough to prevent this, this may be difficult to ensure in practice, especially if multiple users are working on the same course.

Reviewer: Nit, redundant mention of "local":

```suggestion
However, for more complex images or more elaborate changes to existing custom images, it is advisable to test the image in a [local PrairieLearn environment](./installing.md#support-for-external-graders-and-workspaces) before pushing it to the production environment. In this scenario, it is important to ensure that the updated version of the image is restricted to the local environment, and that the production environment is not updated with the new image version until testing is complete. While avoiding an image sync (step 8 above) is often enough to prevent this, this may be difficult to ensure in practice, especially if multiple users are working on the same course.
```

   docker build --platform linux/amd64 -t yourdockerhubaccount/yourimagename .
   ```

   !!! warning

   Note that the production environment used by PrairieLearn uses a `linux/amd64` platform for images. As such it is necessary that images are built using this platform, as listed in `--platform` argument in the command above. This is particularly relevant for users of ARM-based architectures such as Apple M-series processors or Windows on Arm systems, since these systems default to `linux/arm64` instead.

6. Once the build process completes, push the image to the Docker Hub registry (again replacing `yourdockerhubaccount` and `yourimagename` as needed):

   ```bash
   docker push yourdockerhubaccount/yourimagename
   ```

7. Change the settings for the question that needs the custom image to use the image name you set above.
8. In your course's "Sync" page, under "Docker images", find the image name above and click on "Sync".

Note that the process above _will not_ cause the image above to be automatically updated if there are changes or additional features to the original image. It is your responsibility to periodically update your custom image by repeating steps 5-8 above.

> ### Changing existing custom images

Author: I think this is not a major issue when creating a new image, since in that case you don't expect production to make references to the image name, so having the pushed version online is not going to affect the functionality. That said, I can see the appeal with testing locally without the need for pushing, even for new images. I'll update it.
